crates/ra_proc_macro/src/process.rs

```diff
@@ -0,0 +1,204 @@
+//! Handle process life-time and message passing for proc-macro client
+
+use crossbeam_channel::{bounded, Receiver, Sender};
+use ra_tt::Subtree;
+
+use crate::msg::{ErrorCode, Message, Request, Response, ResponseError};
+use crate::rpc::{ExpansionResult, ExpansionTask, ListMacrosResult, ListMacrosTask, ProcMacroKind};
+
+use io::{BufRead, BufReader};
+use std::{
+    io::{self, Write},
+    path::{Path, PathBuf},
+    process::{Child, Command, Stdio},
+    thread::spawn,
+};
+
+#[derive(Debug, Default)]
+pub(crate) struct ProcMacroProcessSrv {
+    inner: Option<Handle>,
+}
+
+struct Task {
+    req: Message,
+    result_tx: Sender<Message>,
+}
+
+#[derive(Debug)]
+struct Handle {
+    sender: Sender<Task>,
+}
+
+struct Process {
+    path: PathBuf,
+    child: Child,
+}
+
+impl Process {
+    fn run(process_path: &Path) -> Result<Process, io::Error> {
+        let child = Command::new(process_path.clone())
+            .stdin(Stdio::piped())
+            .stdout(Stdio::piped())
+            .spawn()?;
+
+        Ok(Process { path: process_path.into(), child })
+    }
+
+    fn restart(&mut self) -> Result<(), io::Error> {
+        let _ = self.child.kill();
+        self.child =
+            Command::new(self.path.clone()).stdin(Stdio::piped()).stdout(Stdio::piped()).spawn()?;
+        Ok(())
+    }
+
+    fn stdio(&mut self) -> Option<(impl Write, impl BufRead)> {
+        let stdin = self.child.stdin.take()?;
+        let stdout = self.child.stdout.take()?;
+        let read = BufReader::new(stdout);
+
+        Some((stdin, read))
+    }
+}
+
+impl ProcMacroProcessSrv {
+    pub fn run(process_path: &Path) -> Result<ProcMacroProcessSrv, io::Error> {
+        let process = Process::run(process_path)?;
+
+        let (task_tx, task_rx) = bounded(0);
+
+        let _ = spawn(move || {
```

Comment:
I think we should save the handle in `ProcMacroProcessSrv` to join the thread in `drop()` as we do it in `ra_cargo_watch::WatchThread`
https://github.com/rust-analyzer/rust-analyzer/blob/d2619bf0ca0c8c2ff8cda22beec9851887e4a4d5/crates/ra_cargo_watch/src/lib.rs#L390-L405

I also wonder if we should generalize this logic of creating a thread to communicate with.
And @matklad might I ask you to elaborate on whether we avoid async/await by-design and the rationale?

Reply:
I improved the shutdown process a little bit. There is a few different between `CargoWatch` and `ProcMacroProcess`, cargo check is a one off process but proc-macro on is a persistent which means we only close the process if RA itself close. Anyway, I added a new [struct](https://github.com/rust-analyzer/rust-analyzer/blob/617f2cde68ba9a2a1974bc7cc0348605f23c6105/crates/ra_proc_macro/src/process.rs#L23) to hold the thread handle for closing the worker thread gracefully.